target/shared/poll.c

```diff
@@ -31,41 +55,111 @@ static int _set_nonblock(int fd)
     return 0;
 }
 
-static void _create_wakefds(void)
+static struct waker* _new_waker(void)
 {
-    if (pipe(_wakefds) == -1)
-        abort();
+    struct waker* ret = NULL;
+    struct waker* waker;
+
+    if (!(waker = malloc(sizeof(struct waker))))
+        goto done;
+
+    waker->next = NULL;
+    waker->tid = _gettid();
+
+    if (pipe(waker->pipefd) == -1)
+        goto done;
+
+    if (_set_nonblock(waker->pipefd[0]) != 0)
+        goto done;
 
-    if (_set_nonblock(_wakefds[0]) != 0)
-        abort();
+    if (_set_nonblock(waker->pipefd[1]) != 0)
+        goto done;
 
-    if (_set_nonblock(_wakefds[1]) != 0)
-        abort();
+    ret = waker;
+    waker = NULL;
+
+done:
+
+    if (waker)
+        free(waker);
+
+    return ret;
 }
 
 long myst_tcall_poll_wake(void)
 {
-    uint64_t x = WAKE_MAGIC;
-
-    /* Create the wake pipe */
-    pthread_once(&_create_wakefds_once, _create_wakefds);
+    /* wake up all waiters */
+    pthread_mutex_lock(&_wakers_mutex);
+    {
+        const uint64_t x = WAKE_MAGIC;
 
-    if (write(_wakefds[1], &x, sizeof(x)) != x)
-        return -EINVAL;
+        for (struct waker* p = _wakers; p; p = p->next)
+        {
+            if (write(p->pipefd[1], &x, sizeof(x)) != sizeof(x))
+            {
+                // return -EINVAL;
```

Comment:
Under which cases would the `write` fail?

Reply:
If the pipe is full (in which case the write is not needed to wake the thread, so can safely be ignored). I will add a comment.